File: src/main/java/me/itzisonn_/meazy/parser/operator/custom/AndOperator.java
package me.itzisonn_.meazy.parser.operator.custom;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.instruction.number.LogicalOperationInstruction.LogicalOperation;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression;
import me.itzisonn_.meazy.parser.operator.Operator;
import me.itzisonn_.meazy.parser.operator.OperatorType;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

import java.lang.constant.ConstantDescs;

@NullMarked
public class AndOperator extends Operator {
    public AndOperator() {
        super("and", "&&", OperatorType.INFIX);
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, OperatorExpression operatorExpression) {
        Expression left = operatorExpression.getLeft();
        Expression right = operatorExpression.getRight();
        if (right == null) throw new NullPointerException("Right side of operator expression is null");

        DataType leftType = left.getType(environment, operatorExpression);
        DataType rightType = right.getType(environment, operatorExpression);
        if (!leftType.getClassDesc().equals(ConstantDescs.CD_boolean) || !rightType.getClassDesc().equals(ConstantDescs.CD_boolean)) throw new RuntimeException("Invalid operands TODO");

        left.emit(instructionsSet, environment, operatorExpression);
        right.emit(instructionsSet, environment, operatorExpression);
        instructionsSet.logicalOperation(LogicalOperation.AND);
    }

    @Override
    public DataType getType(Environment environment, OperatorExpression operatorExpression) {
        return DataType.ofNonNull(ConstantDescs.CD_boolean);
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/operator/custom/AndOperator.kt
package me.itzisonn_.meazy.parser.operator.custom

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.instruction.number.LogicalOperationInstruction.LogicalOperation
import me.itzisonn_.meazy.parser.DataType
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression
import me.itzisonn_.meazy.parser.operator.Operator
import me.itzisonn_.meazy.parser.operator.OperatorType
import me.itzisonn_.meazy.runtime.environment.Environment
import me.itzisonn_.meazy.instruction.isBoolean
import java.lang.constant.ConstantDescs

class AndOperator : Operator("and", "&&", OperatorType.INFIX) {
    override fun emit(instructions: InstructionsSet, environment: Environment, operatorExpression: OperatorExpression) {
        val left = operatorExpression.left
        val right = operatorExpression.right ?: error("Right side of operator expression is null")

        val leftType = left.getType(environment, operatorExpression)
        val rightType = right.getType(environment, operatorExpression)

        if (!leftType.classDesc.isBoolean || !rightType.classDesc.isBoolean) error("Invalid operands TODO")
        if (leftType.isNullable || rightType.isNullable) error("Can't get logical and of nullable booleans")

        left.emit(instructions, environment, operatorExpression)
        instructions.convertToBooleanType(leftType.classDesc == ConstantDescs.CD_Boolean, false)

        right.emit(instructions, environment, operatorExpression)
        instructions.convertToBooleanType(rightType.classDesc == ConstantDescs.CD_Boolean, false)

        instructions.logicalOperation(LogicalOperation.AND)
    }

    override fun getType(environment: Environment, operatorExpression: OperatorExpression): DataType {
        return DataType.ofNonNull(ConstantDescs.CD_boolean)
    }
}
